src/python/pants/backend/python/subsystems/python_tool_base.py

```diff
@@ -127,6 +134,7 @@ def pex_requirements(
         self,
         *,
         extra_requirements: Iterable[str] = (),
+        uses_source_plugins: bool = False,
```

Comment:
You can simplify things by chaing `extra_requirements` to be `requirements_from_source_plugins: Iterable[str] | None = None`. If `requirements_from_source_plugins is None`, assume source plugins are not used. 

Note that both MyPy and Pylint will pass an empty `FrozenOrderedSet`, rather than `None`, if no plugins are used:

https://github.com/pantsbuild/pants/blob/cd24eba1d76c0eb162228666cf27c50121272618/src/python/pants/backend/python/typecheck/mypy/subsystem.py#L219-L221

https://github.com/pantsbuild/pants/blob/cd24eba1d76c0eb162228666cf27c50121272618/src/python/pants/backend/python/typecheck/mypy/subsystem.py#L288-L291

Reply:
Firstly, that line is stale, and I'll be deleting it shortly. I'm not sure that relying on this parameter being set to determine overall tool behaviour is a good idea, because the impact to 3rd-party plugin developers of whether they use `None` or a falsey iterator is pretty obscure at the moment. I've added a field at the class level, which is what we're actually using to implement the behaviour, which keeps things explicit.